rest/route/host_agent_test.go

```diff
@@ -857,295 +858,674 @@ func TestTaskLifecycleEndpoints(t *testing.T) {
+			nextTaskId := "task4"
+			details := &apimodels.GetNextTaskDetails{}
+			task, shouldTeardown, err := assignNextAvailableTask(ctx, env, d.Tq3, model.NewTaskDispatchService(time.Minute), d.Host5, details)
+			require.NoError(t, err)
+			require.NotNil(t, task)
+			assert.False(t, shouldTeardown)
+			assert.Equal(t, nextTaskId, task.Id)
+
+			tq, err := model.LoadTaskQueue(d.Distro3.Id)
+			require.NoError(t, err)
+			assert.Equal(t, 0, tq.Length())
+
+			h, err := host.FindOneId(ctx, d.Host5.Id)
+			require.NoError(t, err)
+			assert.Equal(t, nextTaskId, h.RunningTask)
+		},
+		"tasks with a project with dispatching disabled should be removed from the queue": func(ctx context.Context, t *testing.T, env *mock.Environment, gd func() data) {
+			// The queue has task3 then task4, task3 is under a disabled project.
+			d := gd()
+			d.Project2.DispatchingDisabled = utility.TruePtr()
+			require.NoError(t, d.Project2.Upsert())
+			nextTaskId := d.Tq3.Queue[1].Id
+			details := &apimodels.GetNextTaskDetails{}
+			task, shouldTeardown, err := assignNextAvailableTask(ctx, env, d.Tq3, model.NewTaskDispatchService(time.Minute), d.Host5, details)
+			require.NoError(t, err)
+			require.NotNil(t, task)
+			assert.False(t, shouldTeardown)
+			assert.Equal(t, nextTaskId, task.Id)
+
+			tq, err := model.LoadTaskQueue(d.Distro3.Id)
+			require.NoError(t, err)
+			assert.Equal(t, 0, tq.Length())
+
+			h, err := host.FindOneId(ctx, d.Host5.Id)
+			require.NoError(t, err)
+			assert.Equal(t, nextTaskId, h.RunningTask)
+		},
+		"a completed task group should return a nil task if no task is available ": func(ctx context.Context, t *testing.T, env *mock.Environment, gd func() data) {
+			d := gd()
+			details := &apimodels.GetNextTaskDetails{
+				TaskGroup: "completed-task-group",
 			}
-			taskQueue.Queue = []model.TaskQueueItem{item}
-			So(taskQueue.Save(), ShouldBeNil)
-			t, shouldTeardown, err := assignNextAvailableTask(ctx, env, taskQueue, model.NewTaskDispatchService(time.Minute), &theHostWhoCanBoastTheMostRoast, details)
-			So(err, ShouldBeNil)
-			So(shouldTeardown, ShouldBeFalse)
-			So(t, ShouldBeNil)
-		})
-		Convey("with a host with a running task", func() {
-			anotherHost := host.Host{
-				Id:          "ahost",
-				RunningTask: "sampleTask",
-				Distro: distro.Distro{
-					Id: d.Id,
-				},
-				Secret: hostSecret,
+			task, shouldTeardown, err := assignNextAvailableTask(ctx, env, d.Tq1, model.NewTaskDispatchService(time.Minute), d.Host1, details)
+			require.NoError(t, err)
+			assert.Nil(t, task)
+			assert.True(t, shouldTeardown)
+
+			tq, err := model.LoadTaskQueue(d.Distro1.Id)
+			require.NoError(t, err)
+			assert.Equal(t, tq.Length(), 2)
+
+			h, err := host.FindOneId(ctx, d.Host1.Id)
+			require.NoError(t, err)
+			assert.Equal(t, "", h.RunningTask)
+		},
+		"a completed task group should return the next available task when available": func(ctx context.Context, t *testing.T, env *mock.Environment, gd func() data) {
+			d := gd()
```

Comment:
It seems like you're calling this function for every test case; can you just call it in the for loop section instead, and pass in the result?

Reply:
(I feel like you should just be able to get rid of the function entirely and just do the setup, like https://github.com/evergreen-ci/evergreen/blob/9cfd7e2291abde30e0306cda7c3d925fd7feee95/model/task_lifecycle_test.go#L143 for example)

Reply:
Super true, let me fix this!